hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/metrics/MetricsContext.java

```diff
@@ -209,4 +177,42 @@ protected long get() {
         }
     }
 
+    /**
+     * Map for storing metrics on a per name basis. Meant to offer fast,
+     * single-threaded access, but also a loosely consistent, thread safe
+     * way to iterate over the content.
+     */
+    private static class MetricsStore {
+
+        private Map<String, AbstractMetric> threadLocalStore;
```

Comment:
we are doubling the memory footprint and complicating the implementation to provide a fast local get. Is this really necessary? does it affect the performance that much to use a single `ConcurrentHashMap` ?

Reply:
The `metric` method is called for each processed item, so with just using the `ConcurrentHashMap`, the performance impact might be too big. At least that's what I think...

Reply:
I was curious and suspected that a non-contended read-only CHM might be very similar to plain HM because non-contended volatile read on x86 is as fast as a non-volatile read:
```
Benchmark     Mode  Cnt   Score   Error  Units
ChmBench.chm  avgt   10  21,687 ± 0,122  ns/op
ChmBench.hm   avgt   10  22,610 ± 0,163  ns/op
```
source: 
```java
@BenchmarkMode(Mode.AverageTime)
@OutputTimeUnit(TimeUnit.NANOSECONDS)
@State(Scope.Thread)
@Fork(value = 1, warmups = 1)
@Warmup(iterations = 10)
@Measurement(iterations = 10)
public class ChmBench {

    private final Map<String, Integer> chm = new ConcurrentHashMap<>();
    private final Map<String, Integer> hm = new HashMap<>();

    {
        chm.put("1", 1);
        chm.put("2", 2);
        chm.put("3", 3);

        hm.putAll(chm);
    }

    @Benchmark
    public void chm() {
        chm.get("3");
        chm.get("1");
        chm.get("2");
    }

    @Benchmark
    public void hm() {
        hm.get("3");
        hm.get("1");
        hm.get("2");
    }
}
```
Maybe another CPU will be different, I don't know.